### ISSUE TITLE
DateTime Overhaul

### ISSUE THREAD

**Issue by jchannon** [NancyFx/Nancy#2427]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of Nancy
- [x] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/NancyFx/Nancy/issues) and [closed](https://github.com/NancyFx/Nancy/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
The way Nancy binds and parses bodies and querystring values differs with DateTime objects  This issue is the parent to child issues and PRs whose aim is to investigate and possibly overhaul the DateTime parsing within Nancy.  

Issues that need investigating are:

- [ ] #2417 
- [ ] #2426

One thing I think would be useful is a configuration option in Nancy that tells it how to handle `DateTime`, `DateTimeOffset` is different and more forgiving.  When a date comes into a request via a querystring or body payload the binder or querystring parsing should read the config option and convert the `DateTime` to DateTimeKind.Local, Utc, Unspecified, whichever is selected in the config option. This might be another item in the SDHP, ie/they dont have to convert dates in their modules for example to the format that they want to use in their application. Thoughts?

**Comment by thecodejunkie** [comment]
## Thoughts on #2417 

We have introduced a breaking change between `1.x` and `2.x` when it comes to the deserialization of JSON. When we replaced the JSON serializer with SimpleJson, we also made a change which means that all `DateTime` and `DateTimeOffset` instances will be deserializes with `DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal` (you can see the [SimpleJson code here](https://github.com/NancyFx/Nancy/blob/master/src/Nancy/Json/Simple/SimpleJson.cs#L1346)) and that results in `DateTime` instances being `DateTimeKind.Utc`. 

Say you now return a model with a `DateTime` that has a `+2` offset, for example 10am in Sweden, which will have something like `1000+2` (because Sweden is currently GMT+2 because of daylight savings time) in the serialized representation. When we deserialize this it will think it needs to adjust that to UTC so it will end up as `0800`. In `1.x` this would not happen because the deserialization assumed local.

The reason for the change is that we want to follow the ethos

> `DateTimeOffset` in, `DateTimeOffset` out

The problem is that it will break existing `1.x` clients and they would have to have all consumers update their date/time handling. The proposal is that we had a configuration option that let's you specify which `DateTimeStyles` that should be used, have them default to `DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal` but would still let older applications change the configuration to use `DateTimeStyles.AdjustToLocal`. The conundrum is how we get that information into the `JavaScriptSerializer`, which is what is used by both `DefaultJsonSerializer` and `JsonBodyDeserializer` (it's actually only the second one that is important since this is a deserialization concern.

### Proposed solution 1 - Add it to `BindingConfig`

I.e the config that can be passed in to `Bind<T>` in a module. This would let us control is on a per-bind basis. However, since will only flow to the `JavaScriptSerializer` via the  `JsonBodyDeserializer`, since binding does not use `DefaultJsonSerializer`. This means we would need to hard code values in the `DefaultJsonSerializer` that is passes to the `JavaScriptSerializer` when it uses it, even though the values aren't going to have any impact. 

The default value would be `DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal`

### Proposed solution 2 - Add a `ModelBindingConfiguration` environment config

This would let us do

```
env.Binding(
   dateTimeStyles: DateTimeStyles.AdjustToLocal);
```

The default value would be `DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal`

Both `DefaultJsonSerializer` and `JsonBodyDeserializer` already have access to the `INancyEnvironment` and already use it to pass configs into the `JavaScriptSerializer`.

However I don't feel this sits quite right because it's not really only related to model binding, but more to JSON (de)serialization in general, which leads me to proposal 3

### Proposed solution 2 - Add option to the `JsonConfiguration` environment config

We already have a `JsonConfiguration` environment config and since  `DefaultJsonSerializer` and `JsonBodyDeserializer` already have access to the `INancyEnvironment` and already use it to pass configs into the `JavaScriptSerializer`, this sounds like a clear winner. It would not let us do per-bind decisions on this, but it would not prevent anyone from doing that conversion themselves if/when they need it.

**Comment by thecodejunkie** [comment]
## Thoughts on #2426 

tl;dr The `DynamicDictionaryValue` does `DateTime` parsing in two places([#1](https://github.com/NancyFx/Nancy/blob/master/src/Nancy/DynamicDictionaryValue.cs#L92) and [#2](https://github.com/NancyFx/Nancy/blob/master/src/Nancy/DynamicDictionaryValue.cs#L418)) and neither assumes UTC. This means that we know have different behavior between `DateTime` parsing in `DynamicDictionaryValue` and how JSON deserialization treats them.

The problem is that we have no, obvious, way of injecting any config values into `DynamicDictionaryValue` to let the user define the expected behavior.

Thoughts on solutions?!

ping @NancyFx/most-valued-minions

**Comment by jchannon** [comment]
#2417 proposed solution 2 makes sense to put it in the JsonConfiguration

#2426 only thing I can thing is to make `Default and TryParse` take the `JsonConfiguration` as a second method argument so users can choose.  If it's not provided we use `JsonConfiguration.Default` in `DynamicDictionaryValue`

**Comment by jchannon** [comment]
Addressed by https://github.com/NancyFx/Nancy/pull/2438/ & https://github.com/NancyFx/Nancy/pull/2454
